lib/improver/between_thresholds.py

```diff
@@ -0,0 +1,205 @@
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Plugin to calculate probabilities of occurrence between specified thresholds
+"""
+
+import iris
+import numpy as np
+
+from iris.exceptions import CoordinateNotFoundError
+
+from improver.utilities.cube_checker import find_threshold_coordinate
+from improver.utilities.cube_metadata import extract_diagnostic_name
+
+
+class OccurrenceBetweenThresholds(object):
+    """Calculate the probability of occurrence between thresholds"""
+
+    def __init__(self, threshold_ranges, threshold_units):
+        """
+        Initialise the class
+
+        Args:
+            threshold_ranges (list):
+                List of 2-item iterables specifying thresholds between which
+                probabilities should be calculated
+            threshold_units (str):
+                Units in which the thresholds are specified
+        """
+        self.threshold_ranges = threshold_ranges
+        self.threshold_units = threshold_units
+
+    def _slice_cube(self, cube):
+        """
+        Extract required slices from input cube
+
+        Args:
+            cube (iris.cube.Cube):
+                Input cube
+
+        Returns:
+            cubes (list):
+                List of 2-item lists containing lower and upper
+                threshold cubes
+
+        Raises:
+            ValueError:
+                If any of the required constraints returns None
+        """
+        thresh_coord = find_threshold_coordinate(cube)
+        error_string = thresh_coord.name() + ' threshold {} is not available\n'
+        error_msg = ''
+
+        cubes = []
+        for t_range in self.threshold_ranges:
+            t_range.sort()
+            lower_constraint = iris.Constraint(coord_values={
+                thresh_coord: lambda t: np.isclose(t.point, t_range[0])})
+            lower_cube = cube.extract(lower_constraint)
+            if lower_cube is None:
+                error_msg += error_string.format(t_range[0])
+            upper_constraint = iris.Constraint(coord_values={
+                thresh_coord: lambda t: np.isclose(t.point, t_range[1])})
```

Comment:
Good point.  Will specify a sensible `atol` to address this.

Reply:
OK, actually I had no issue with this - the unit test I wrote for this passed straight away.  But if it doesn't work on all hardware, I'll try setting `atol` to `10^-6` for this case.

Reply:
Ah actually I'm not going to unless I need to, for the reason that if we use this for precipitation rates in `m s-1`, this will be a problem.

Reply:
Your test passes because you have not made the coordinate in your unit test of type float32. If you change it as below you will find that the test fails and you still need to handle this.
```
self.temp_cube.coord('air_temperature').points = np.array([
    272.15, 273.15, 274.15, 275.15], dtype=np.float32)
```
Good old float32.

Reply:
Eurgh.  This requires an absolute tolerance of `1e-5`.  I'll make a note in the docstring.

Reply:
It's not lovely. I'm struggling to think of much else we can do without making horrible special cases. The horrible special cases treatment might be something like adding 1 to the desired threshold and coordinate when the abs(threshold) < 1.0E-4. This would allow rtol to play a role in slightly expanding the tolerance without having to blanket raise atol, but this is a horrible fudge fix.

End of my musings. I think what you've done will have to suffice for now.